pkg/atconf/file_writer.go

```diff
@@ -0,0 +1,339 @@
+	} else {
+		flagVal = "False"
+	}
+	return f.Cfg.Set(ConfigurationSection, ConfigurationIPv6Option, flagVal)
+}
+
+// addNewHosts adds the pods as new hosts to the admintools.conf file.  It works
+// on admintools.conf using the in-memory ConfigParser representation.
+func (f *FileWriter) addNewHosts(installIPs []string) error {
+	oldHosts := f.getHosts()
+	if err := f.addToClusterHosts(oldHosts, installIPs); err != nil {
+		return err
+	}
+	return f.addNodes(oldHosts, installIPs)
+}
+
+// removeOldHosts will remove the given IPs from the admintools.conf.  Changes
+// are made in-place in ConfigParser.
+func (f *FileWriter) removeOldHosts(ips []string) error {
+	if err := f.removeFromClusterHosts(ips); err != nil {
+		return err
+	}
+	return f.removeNodes(ips)
+}
+
+// addToClusterHosts will add the given set of installIPs as new hosts to the
+// Cluster section.  The updates are done in-place in the ConfigParser.
+func (f *FileWriter) addToClusterHosts(oldHosts map[string]bool, installIPs []string) error {
+	var ips strings.Builder
+	oldHostLine, err := f.Cfg.Get(ClusterSection, ClusterHostOption)
+	// Ignore error in case the hosts option doesn't exist
+	if err == nil {
+		ips.WriteString(oldHostLine)
+	}
+	for _, ip := range installIPs {
+		// If host already exists, we treat as a no-op and skip the host
+		if _, ok := oldHosts[ip]; ok {
+			continue
+		}
+		if ips.Len() != 0 {
+			ips.WriteString(",")
+		}
+		ips.WriteString(ip)
+	}
+	err = f.Cfg.Set(ClusterSection, ClusterHostOption, ips.String())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// removeFromClusteHosts will remove a set of IPs from the Cluster.hosts section
+// of the config.  Changes are made in-place in the ConfigParser.
+func (f *FileWriter) removeFromClusterHosts(ips []string) error {
+	oldHostLine, err := f.Cfg.Get(ClusterSection, ClusterHostOption)
+	if err != nil {
+		return err
+	}
+	hosts := strings.Split(oldHostLine, ",")
+	for _, removeIP := range ips {
+		for i := len(hosts) - 1; i >= 0; i-- {
+			if hosts[i] == removeIP {
+				hosts = append(hosts[0:i], hosts[i+1:]...)
+				break
+			}
+		}
+	}
+	return f.Cfg.Set(ClusterSection, ClusterHostOption, strings.Join(hosts, ","))
+}
+
+// addNodes will add the given set of installIPs as new nodes in the Nodes
+// section.  The updates are done in-place in the ConfigParser.
+func (f *FileWriter) addNodes(oldHosts map[string]bool, installIPs []string) error {
+	nextNodeNumber := f.getNextNodeNumber()
+	for _, ip := range installIPs {
+		// If host already exists, we treat as a no-op and skip the host
+		if _, ok := oldHosts[ip]; ok {
+			continue
+		}
+		nodeName := fmt.Sprintf("node%04d", nextNodeNumber)
+		nextNodeNumber++
+		nodeInfo := fmt.Sprintf("%s,%s,%s", ip, f.Vdb.Spec.Local.DataPath, f.Vdb.Spec.Local.DataPath)
+		err := f.Cfg.Set(NodesSection, nodeName, nodeInfo)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// removeNodes will remove the nodes section for the given set of IPs
```

Comment:
I was wondering if this function is expected to remove only compat21 node entries, if yes, should we assert or check somewhere to make sure we do not touch regular database nodes, at least mention in the comment? I understand those regular database nodes will be removed and are supposed to be removed by db_remove_subcluster/db_remove_nodes before the uninstallation.

Reply:
That's right.  We only remove compat21 node entries.  The AT -t db_remove_nodes will handle removing of regular database nodes.  I will mention in a comment.